sycl/include/CL/sycl/stream.hpp

```diff
@@ -236,6 +224,18 @@ class stream {
     detail::workGroupBarrier();
     WIOffset = WIOffsetAcc[0].fetch_add(FlushBufferSize);
   }
+
+  void __finalize() {
+    // Flush data to global buffer in stream destruction if flush buffer is not
```

Comment:
This happens not during destruction now, isn't?

Reply:
Yes, it is not in destructor, now. There is no user-defined destructor, now.

A call to __finalize method is now substituted at the kernel end.
Some use cases do not allow to pass pointer to object to kernel without byval attribute. 
Having a destructor in some class prohibits to pass pointer to its instance with byval attribute.
Regarding that cl::sycl::stream class is our property we are free to let it have __finalize. It already has an __init method and we're OK with that.

Reply:
I know where this whole patch comes from.
But, generation of  `__init` method call is ugly but necessary solution, it allows to deal with OpenCL limitations and lower SYCL kernel object to OpenCL-like kernels (see https://github.com/intel/llvm/blob/sycl/sycl/doc/SYCLCompilerAndRuntimeDesign.md#lowering-of-lambda-function-objects-and-named-function-objects).
The fact that we need some `__finalize` call for stream is our implementation detail. I feel there should be way to resolve this problem in SYCL headers instead of extending ugliness.

Reply:
Currently the stream is flushed on either `cl::sycl::stream_manipulator::endl` or `cl::sycl::stream_manipulator::flush`. The spec requires that instance of `cl::sycl::stream` class in sychronized with the host and output everything that is stream to it via `operator<<()` by the time that the kernel function finishes execution. 
Thus, we can't leave buffers unflushed upon kernel finish.
On the other hand we can't flush buffer upon each and every call to `operator<<()` which will otherwise cause an incomprehensible mess in output of stream.

Reply:
@Fznamznon 
> I know where this whole patch comes from.
But, generation of  `__init` method call is ugly but necessary solution, it allows to deal with OpenCL limitations and lower SYCL kernel object to OpenCL-like kernels (see https://github.com/intel/llvm/blob/sycl/sycl/doc/SYCLCompilerAndRuntimeDesign.md#lowering-of-lambda-function-objects-and-named-function-objects).
The fact that we need some `__finalize` call for stream is detail of our stream class implementation. I feel there should be way to resolve this problem in SYCL headers instead of extending ugliness.

We can't add call to __finalize to kernel_singe_task et al because it is replaced at code-gen step (or somewhere about it). AFAIK currently command group is only released when there is no more buffers associated with it. The stream class contains its buffers whilst it is created inside command group. C-tor of the `cl::sycl::stream` accepts ref to command group handler by spec. All this make it not very possible to resolve this issue within SYCL runtime. They say. there is big feature missing in scheduler: release of each and every command group regardless of buffers' state.

@againull and @romanovvlad  may have concerns here

Reply:
@Fznamznon 
>>  I feel there should be way to resolve this problem in SYCL headers instead of extending ugliness.

Could you please be more specific, what alternative solution do you propose?

Reply:
> @Fznamznon
> 
> > > I feel there should be way to resolve this problem in SYCL headers instead of extending ugliness.
> 
> Could you please be more specific, what alternative solution do you propose?

Discussed offline with Sergey. I suggest to extend functions which we use to generate kernels. But, I'm not sure that it will work.
But if it will, we can try to avoid huge change in clang and move OpenCL kernel generation to SYCL headers.

Reply:
@Fznamznon The change is already in method CreateOpenCLKernelBody.

Reply:
@s-kanaev , I meant extending functions with `sycl_kernel` attribute which are implemeted in SYCL headers, not the compiler part.
BTW, after some thinking, I don't think that it's possible to do.

Please remove mention about destruction from comment in `__finalize` method and add NOTE that usage of this method will be added by the compiler.

Reply:
> Please remove mention about destruction from comment in __finalize method and add NOTE that usage of this method will be added by the compiler.


Done